# Homework 1

### 1. Taking inspiration from the notebook `01-intro-to-pt.ipynb`, build a class for the Multilayer Perceptron (MLP) whose scheme is drawn in the last figure of the notebook. As written there, no layer should have bias units and the activation for each hidden layer should be the Rectified Linear Unit (ReLU) function, also called ramp function. The activation leading to the output layer, instead, should be the softmax function, which prof. Ansuini explained during the last lecture. You can find some notions on it also on the notebook.                    


In [8]:
import torch
import numpy as np

In [13]:
class MLP(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.layer1 = torch.nn.Linear(in_features=5, out_features=11, bias = False)
        self.layer2 = torch.nn.Linear(in_features=11, out_features=16, bias = False)
        self.layer3 = torch.nn.Linear(in_features=16, out_features=13, bias = False)
        self.layer4 = torch.nn.Linear(in_features=13, out_features=8, bias = False)
        self.layer5 = torch.nn.Linear(in_features=8, out_features=4, bias = False)

    def forward(self, X):
        out = self.layer1(X)
        out = torch.nn.functional.relu(out)
        out = self.layer2(out)
        out = torch.nn.functional.relu(out)
        out = self.layer3(out)
        out = torch.nn.functional.relu(out)
        out = self.layer4(out)
        out = torch.nn.functional.relu(out)
        out = self.layer5(out)
        out = torch.nn.functional.softmax(out)
        return out


### 2. After having defined the class, create an instance of it and print a summary using a method of your choice.


In [14]:
import sys
!{sys.executable} -m pip install torch-summary #how to use pip or conda in jupyter notebooks
from torchsummary import summary

In [39]:
from random import seed 
seed(123456)
model = MLP()
summary(model)

Layer (type:depth-idx)                   Param #
├─Linear: 1-1                            55
├─Linear: 1-2                            176
├─Linear: 1-3                            208
├─Linear: 1-4                            104
├─Linear: 1-5                            32
Total params: 575
Trainable params: 575
Non-trainable params: 0


Layer (type:depth-idx)                   Param #
├─Linear: 1-1                            55
├─Linear: 1-2                            176
├─Linear: 1-3                            208
├─Linear: 1-4                            104
├─Linear: 1-5                            32
Total params: 575
Trainable params: 575
Non-trainable params: 0


### 3. Provide detailed calculations (layer-by-layer) on the exact number of parameters in the network.





A set of parentheses for each layer, without bias units:
$$(5 \times 11)+(11 \times 16) + (16 \times 13) + (13 \times 8) + (8 \times 4) = 55 + 176 + 208 + 104 + 32 = 575$$

### Provide the same calculation in the case that the bias units are present in all layers (except input).

A set of parentheses for each layer, in each the additive term represents the biases.

$$(5 \times 11+11)+(11 \times 16+16) + (16 \times 13 +13) + (13 \times 8+8) + (8 \times 4 +4) = 66 + 192 + 221 + 112 + 36 = 627$$

### 4. For each layer within the MLP, calculate the L2 norm and L1 norm of its parameters.

In [46]:
# L2: Sum of the squared weights
# L1: Sum of the absolute weights

i = 0
for param in model.parameters():
  print("LAYER", i)
  print("Norm L1:", torch.norm(param, 1).item())
  print("Norm L2:", torch.norm(param, 2).item())
  print("")
  i+=1

LAYER 0
Norm L1: 13.130066871643066
Norm L2: 2.0066583156585693

LAYER 1
Norm L1: 27.10474395751953
Norm L2: 2.398174524307251

LAYER 2
Norm L1: 26.631126403808594
Norm L2: 2.150670289993286

LAYER 3
Norm L1: 15.996356964111328
Norm L2: 1.7539299726486206

LAYER 4
Norm L1: 5.739292621612549
Norm L2: 1.175723671913147

